synapse/lib/stormtypes.py

```diff
@@ -122,23 +122,120 @@ def _validateInfo(self, obj, info, name):
+                    rname.remove('stor')
+                if isinstance(rname, dict):
+                    rname.pop('stor', None)
+                isstor = True
+
+            invalid = (isgtor and isctor) or (isfunc and (isgtor or isctor))
+            if invalid:
+                mesg = f'Dictionary represents invalid combination of ctors, gtors, locls, and stors [{name} {obj} {info.get("name")}] [{rtype}].'
+                raise AssertionError(mesg)
+
+            if rname:
+                mesg = f'Dictionary return types represents a unknown rtype [{name} {obj} {info.get("name")}] [{rtype}] [{rname}].'
+                raise AssertionError(mesg)
+
+            if isfunc:
+                self._validateFunction(obj, info, name)
+            if isstor:
+                self._validateStor(obj, info, name)
+            if isctor:
+                self._validateCtor(obj, info, name)
+            if isgtor:
+                self._validateGtor(obj, info, name)
+
+    def _validateFunction(self, obj, info, name):
+        rtype = info.get('type')
+        funcname = rtype.pop('_funcname')
+        locl = getattr(obj, funcname, None)
+        assert locl is not None, f'bad funcname=[{funcname}] for {obj} {info.get("name")}'
+        args = rtype.get('args', ())
+        callsig = getCallSig(locl)
+        # Assert the callsigs match
+        callsig_args = [str(v).split('=')[0] for v in callsig.parameters.values()]
+        assert [d.get('name') for d in
+                args] == callsig_args, f'args / callsig args mismatch for {funcname} {name} {obj} {args} {callsig_args}'
+        # ensure default values are provided
+        for parameter, argdef in zip(callsig.parameters.values(), args):
+            pdef = parameter.default  # defaults to inspect._empty for undefined default values.
+            adef = argdef.get('default', inspect._empty)
+            assert pdef == adef, \
+                f'Default value mismatch for {obj} {funcname}, defvals {pdef} != {adef} for {parameter}'
+
+    def _validateStor(self, obj, info, name):
+        rtype = info.get('type')
+        funcname = rtype.pop('_storfunc')
+        locl = getattr(obj, funcname, None)
+        assert locl is not None, f'bad funcname=[{funcname}] for {obj} {info.get("name")}'
+        args = rtype.get('args')
+        assert args is None, f'stors have no defined args funcname=[{funcname}] for {obj} {info.get("name")}'
+        callsig = getCallSig(locl)
+        # Assert the callsig for a stor has one argument
+        callsig_args = [str(v).split('=')[0] for v in callsig.parameters.values()]
+        assert len(callsig_args) == 1, f'stor funcs must only have one argument for {obj} {info.get("name")}'
+
+    def _validateCtor(self, obj, info, name):
+        rtype = info.get('type')
+        funcname = rtype.pop('_ctorfunc')
+        locl = getattr(obj, funcname, None)
+        assert locl is not None, f'bad funcname=[{funcname}] for {obj} {info.get("name")}'
```

Comment:
Could we change this and the `_validateStor/Gtor` ones to start like `bad <ctor/gtor/stor>name` so it's more immediately apparent which failed?

Reply:
yup!